futures/tests/io_buf_writer.rs

```diff
@@ -0,0 +1,247 @@
+    assert_eq!(&writer.get_ref().inner, &[0, 1]);
+
+    run_write!(&mut writer, &[3]).unwrap();
+    assert_eq!(writer.buffer(), [2, 3]);
+    assert_eq!(&writer.get_ref().inner, &[0, 1]);
+
+    run(writer.flush()).unwrap();
+    assert_eq!(writer.buffer(), []);
+    assert_eq!(&writer.get_ref().inner, &[0, 1, 2, 3]);
+
+    run_write!(&mut writer, &[4]).unwrap();
+    run_write!(&mut writer, &[5]).unwrap();
+    assert_eq!(writer.buffer(), [4, 5]);
+    assert_eq!(&writer.get_ref().inner, &[0, 1, 2, 3]);
+
+    run_write!(&mut writer, &[6]).unwrap();
+    assert_eq!(writer.buffer(), [6]);
+    assert_eq!(writer.get_ref().inner, &[0, 1, 2, 3, 4, 5]);
+
+    run_write!(&mut writer, &[7, 8]).unwrap();
+    assert_eq!(writer.buffer(), []);
+    assert_eq!(writer.get_ref().inner, &[0, 1, 2, 3, 4, 5, 6, 7, 8]);
+
+    run_write!(&mut writer, &[9, 10, 11]).unwrap();
+    assert_eq!(writer.buffer(), []);
+    assert_eq!(writer.get_ref().inner, &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
+
+    run(writer.flush()).unwrap();
+    assert_eq!(writer.buffer(), []);
+    assert_eq!(&writer.get_ref().inner, &[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
+}
+
+#[test]
+fn maybe_pending_buf_writer_inner_flushes() {
+    let mut w = BufWriter::with_capacity(3, MaybePending::new(Vec::new()));
+    run_write!(&mut w, &[0, 1]).unwrap();
+    assert_eq!(&w.get_ref().inner, &[]);
+    run(w.flush()).unwrap();
+    let w = w.into_inner().inner;
+    assert_eq!(w, [0, 1]);
+}
+
+
+struct MaybePendingSeek {
+    inner: Cursor<Vec<u8>>,
+    ready_write: bool,
+    ready_seek: bool,
+}
+
+impl MaybePendingSeek {
+    fn new(inner: Vec<u8>) -> Self {
+        Self { inner: Cursor::new(inner), ready_write: false, ready_seek: false }
+    }
+}
+
+impl AsyncWrite for MaybePendingSeek {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        if self.ready_write {
+            self.ready_write = false;
+            Pin::new(&mut self.inner).poll_write(cx, buf)
+        } else {
+            self.ready_write = true;
+            Poll::Pending
+        }
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Pin::new(&mut self.inner).poll_flush(cx)
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        Pin::new(&mut self.inner).poll_close(cx)
+    }
+}
+
+impl AsyncSeek for MaybePendingSeek {
+    fn poll_seek(mut self: Pin<&mut Self>, cx: &mut Context<'_>, pos: SeekFrom)
+        -> Poll<io::Result<u64>>
+    {
+        if self.ready_seek {
+            self.ready_seek = false;
+            Pin::new(&mut self.inner).poll_seek(cx, pos)
+        } else {
+            self.ready_seek = true;
+            Poll::Pending
+        }
+    }
+}
+
+#[test]
+fn maybe_pending_buf_writer_seek() {
+    // FIXME: when https://github.com/rust-lang-nursery/futures-rs/issues/1510 fixed,
```

Comment:
Now that `AsyncSeek` has been merged, this can probably be fixed, right?

Reply:
As far as I tried locally, it still didn't work...

Reply:
Ah, yeah, the choice currently is to have a `Vec<u8>` which is `AsyncWrite` and extends as needed or `Cursor<Vec<u8>>` which is `AsyncWrite + AsyncSeek` but doesn't extend as needed #1510 is about whether we can have a type that implements both without having to drop the genericness of the current `Cursor` implementation.

Reply:
I threw together a test commit and this can definitely be done via specialization: https://github.com/rust-lang-nursery/futures-rs/commit/7997066642e5120db44f9eb7ceac612ca7411316, if only it were close to stabilization.